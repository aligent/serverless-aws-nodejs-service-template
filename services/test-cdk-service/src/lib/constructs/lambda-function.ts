import { Duration, Stack } from 'aws-cdk-lib';
import { Runtime, Tracing } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction, type NodejsFunctionProps } from 'aws-cdk-lib/aws-lambda-nodejs';

export interface LambdaFunctionProps extends NodejsFunctionProps {
    readonly entry: string;
}

export class LambdaFunction extends NodejsFunction {
    constructor(scope: Stack, id: string, props: LambdaFunctionProps) {
        super(scope, id, {
            functionName: `${scope.stackName}-${id}`,
            timeout: Duration.seconds(6),
            memorySize: 192,
            // TODO: Make this reference a global attribute across the whole app and default to LTS if not found
            runtime: Runtime.NODEJS_20_X,
            tracing: Tracing.ACTIVE,
            // TODO: This may need to be merged more carefully
            environment: {
                NODE_OPTIONS: '--enable-source-maps',
            },
            bundling: {
                sourceMap: true,
            },
            ...props,
        });

        this.addAlias('LATEST');
    }
}
