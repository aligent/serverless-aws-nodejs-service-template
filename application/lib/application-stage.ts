import { CdkServiceStack } from '@services/cdk-service';
import { Stage, Tags, type StageProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';

/**
 * StageId can be any string, but it's strongly recommended to use
 * a 3 letter preset. This type uses a trick to provide recommended
 * options in intellisense while accepting any string.
 */
type StageId = 'dev' | 'stg' | 'prd' | (string & {});

/**
 * ApplicationStage is the main repeatable unit of deployment.
 * It should contain all the stacks required by this application
 */
export class ApplicationStage extends Stage {
    constructor(scope: Construct, stage: StageId, props?: StageProps) {
        super(scope, stage as string, props);

        new CdkServiceStack(this, 'cdk-service', {
            ...props,
            description: 'CDK service template generated using Nx',
        });

        Tags.of(this).add('STAGE', stage);
    }
}
