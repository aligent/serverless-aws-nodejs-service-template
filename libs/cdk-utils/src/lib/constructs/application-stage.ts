import { Aspects, Stage, Tags, type StageProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { FunctionBundleAnalyzerAspect } from '../aspects/function-bundle-analyzer';
import { LambdaFunction } from './lambda-function';

abstract class BaseApplicationStage extends Stage {
    static readonly DEFAULT_STAGES = {
        development: 'dev',
        staging: 'stg',
        production: 'prd',
    } as const;
}

type DefaultStageOptions = typeof BaseApplicationStage.DEFAULT_STAGES;
type DefaultStageId = DefaultStageOptions[keyof DefaultStageOptions];

/**
 * Constructor for stages with a custom id.
 *
 * Use of standard stage ids will throw an error.
 */
export class CustomApplicationStage extends BaseApplicationStage {
    constructor(scope: Construct, stage: string, props?: StageProps) {
        if (Object.values(BaseApplicationStage.DEFAULT_STAGES).includes(stage as DefaultStageId)) {
            throw new Error(
                `Stage ID ${stage} is reserved for default stages, use the appropriate constructor`
            );
        }
        super(scope, stage, props);
    }
}

/**
 * Development ApplicationStage
 *
 * Adds an appropriate stage tag.
 * Adds the FunctionBundleAnalyzerAspect.
 */
export class DevelopmentApplicationStage extends BaseApplicationStage {
    constructor(scope: Construct, props?: StageProps) {
        super(scope, BaseApplicationStage.DEFAULT_STAGES.development, props);

        Tags.of(this).add('STAGE', BaseApplicationStage.DEFAULT_STAGES.development);

        // This aspect enables using the `analyze:function` npm script to
        // display bundle analyses for lambda functions
        Aspects.of(this).add(
            new FunctionBundleAnalyzerAspect({
                customFunctionConstruct: LambdaFunction,
            })
        );
    }
}

/**
 * Staging ApplicationStage
 *
 * Adds an appropriate stage tag.
 */
export class StagingApplicationStage extends BaseApplicationStage {
    constructor(scope: Construct, props?: StageProps) {
        super(scope, BaseApplicationStage.DEFAULT_STAGES.staging, props);

        Tags.of(this).add('STAGE', BaseApplicationStage.DEFAULT_STAGES.staging);
    }
}

/**
 * Production ApplicationStage
 *
 * Adds an appropriate stage tag.
 * Constructs can apply production-specific configuration by checking for this stage
 *
 * @example
 * ```ts
 * const stage = Stage.of(scope);
 *
 * if (stage instanceof ProductionApplicationStage) {
 *     // Apply production-specific configuration
 * }
 * ```
 */
export class ProductionApplicationStage extends BaseApplicationStage {
    constructor(scope: Construct, props?: StageProps) {
        super(scope, BaseApplicationStage.DEFAULT_STAGES.production, props);

        Tags.of(this).add('STAGE', BaseApplicationStage.DEFAULT_STAGES.production);
    }
}
