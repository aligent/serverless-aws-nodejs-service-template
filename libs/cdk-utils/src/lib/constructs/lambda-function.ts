import { RemovalPolicy, Stage } from 'aws-cdk-lib';
import {
    LogLevel,
    NodejsFunction,
    OutputFormat,
    type NodejsFunctionProps,
} from 'aws-cdk-lib/aws-lambda-nodejs';
import { LogGroup, RetentionDays } from 'aws-cdk-lib/aws-logs';
import { propertyInjectable } from 'aws-cdk-lib/core/lib/prop-injectable';
import { Construct } from 'constructs';
import { DevelopmentApplicationStage, StagingApplicationStage } from './application-stage';

/**
 * LambdaFunctionProps
 *
 * @see https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_lambda_nodejs.NodejsFunctionProps.html
 * for properties inherited from NodejsFunction
 */
export interface LambdaFunctionProps extends NodejsFunctionProps {
    /**
     * The entry point for the Lambda function.
     * This should be the fully resolved path to the handler file.
     *
     * @example path.join(__dirname, 'cdk/handlers/remap-eta.ts')
     */
    readonly entry: string;
    /**
     * Alias for the Lambda function.
     * If undefined, the function will not be versioned or aliased.
     *
     * @example 'LATEST'
     */
    readonly alias?: string;
}

/**
 * Narrowed type helper for setting global defaults
 *
 * Exclusions:
 * logGroup, entry point - doesn't make sense to set on a global level
 * logRetention - legacy API
 *
 * @see https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_lambda_nodejs.NodejsFunctionProps.html
 * for properties inherited from NodejsFunction
 */
type LambdaFunctionContext = Omit<LambdaFunctionProps, 'entry' | 'logGroup' | 'logRetention'>;

/**
 * LambdaFunction construct
 *
 * Set application-level defaults using the context api
 *
 * @example
 * ```ts
 * const APPLICATION_CONTEXT = {
 *     ...LambdaFunction.defineContext({
 *         runtime: Runtime.NODEJS_20_X,
 *     }),
 * };
 * ```
 *
 * @see https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_lambda_nodejs.NodejsFunctionProps.html
 * for properties inherited from NodejsFunction
 */
@propertyInjectable
export class LambdaFunction extends NodejsFunction {
    readonly PROPERTY_INJECTION_ID = '@aligent.cdk-utils.LambdaFunction';
    static readonly CONTEXT_KEY = '@aligent.cdk-utils.LambdaFunction';

    /**
     * Define context properties for the LambdaFunction construct
     *
     * @param props - The properties for the LambdaFunction construct
     * @returns The context for the LambdaFunction construct
     *
     * @see https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_lambda_nodejs.NodejsFunctionProps.html
     * for properties inherited from NodejsFunction
     */
    static defineContext(props: LambdaFunctionContext) {
        return {
            [this.CONTEXT_KEY]: props,
        };
    }

    constructor(scope: Construct, id: string, props: LambdaFunctionProps) {
        // Get base defaults from context
        const defaults: LambdaFunctionContext =
            scope.node.tryGetContext(LambdaFunction.CONTEXT_KEY) || {};

        // Get the current stage
        const currentStage = Stage.of(scope);
        if (!currentStage) {
            throw new Error('This construct must be used within a CDK Stage');
        }

        // Create a log group for the function with stage-appropriate properties if one hasn't been provided
        const logGroup =
            props.logGroup ||
            new LogGroup(scope, `/aws/lambda/${id}`, logGroupProperties(currentStage));

        // Apply stage-specific bundling optimizations
        const bundling = {
            ...bundlingProperties(currentStage),
            ...defaults.bundling,
            ...props.bundling,
        };

        const environment = {
            ...defaults.environment,
            ...props.environment,
            // Always enable source maps in the environment if they've been set for bundling
            ...(bundling.sourceMap ? { NODE_OPTIONS: '--enable-source-maps' } : {}),
        };

        // Merge all configurations with stage-optimized bundling taking precedence
        // TODO may need to do a proper deep merge here
        const finalProps = {
            ...defaults,
            ...props,
            environment,
            bundling,
            logGroup,
        };

        super(scope, id, finalProps);

        if (props.alias) {
            this.addAlias(props.alias);
        }
    }
}

/**
 * Get stage-specific log group properties
 *
 * Development and staging log groups don't need to last forever
 * and should be cleaned up along with their stack when removed
 *
 * @param stage - The stage to get the log group properties for
 * @returns The log group properties for the stage
 */
function logGroupProperties(stage: Stage) {
    const devConfig = {
        retention: RetentionDays.ONE_WEEK,
        removalPolicy: RemovalPolicy.DESTROY,
    };

    const stagingConfig = {
        retention: RetentionDays.SIX_MONTHS,
        removalPolicy: RemovalPolicy.DESTROY,
    };

    const productionConfig = {
        retention: RetentionDays.TWO_YEARS,
        removalPolicy: RemovalPolicy.RETAIN,
    };

    switch (true) {
        case stage instanceof DevelopmentApplicationStage:
            return devConfig;
        case stage instanceof StagingApplicationStage:
            return stagingConfig;
        default:
            return productionConfig;
    }
}

/**
 * Get stage-specific bundling configuration
 *
 * Production stages use optimized settings for performance
 * Development stages use development-friendly settings
 *
 * @param stage - The stage to get the bundling properties for
 * @returns The bundling properties for the stage
 */
function bundlingProperties(stage: Stage) {
    // Dev configuration optimised for debugging, analysis, and bundling speed
    const devConfig = {
        logLevel: LogLevel.INFO,
        minify: false,
        sourceMap: false,
        metafile: true, // Required for bundle analysis
        format: OutputFormat.ESM,
    } satisfies NodejsFunctionProps['bundling'];

    // Staging/Production configurations optimised for bundle size
    const stagingConfig = {
        keepNames: true,
        logLevel: LogLevel.INFO,
        minify: true,
        sourceMap: true,
        buildArgs: {
            bundle: 'true',
            treeShaking: 'true',
        },
        format: OutputFormat.ESM,
    } satisfies NodejsFunctionProps['bundling'];

    const productionConfig = {
        keepNames: true,
        // TODO is this necessary? It might only apply to ESBuild logs
        logLevel: LogLevel.INFO,
        minify: true,
        // TODO: Source maps are really large, consider disabling for production
        sourceMap: true,
        buildArgs: {
            bundle: 'true',
            treeShaking: 'true',
        },
        format: OutputFormat.ESM,
    } satisfies NodejsFunctionProps['bundling'];

    switch (true) {
        case stage instanceof DevelopmentApplicationStage:
            return devConfig;
        case stage instanceof StagingApplicationStage:
            return stagingConfig;
        default:
            return productionConfig;
    }
}
