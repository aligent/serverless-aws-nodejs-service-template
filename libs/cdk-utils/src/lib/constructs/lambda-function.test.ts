import { App, Stack } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import { LogGroup } from 'aws-cdk-lib/aws-logs';
import path from 'path';
import { beforeEach, describe, expect, it, test } from 'vitest';
import { DevelopmentApplicationStage, StagingApplicationStage } from './application-stage';
import { LambdaFunction } from './lambda-function';

describe('LambdaFunction', () => {
    describe('unit tests', () => {
        it('should have the correct static context key', () => {
            expect(LambdaFunction.CONTEXT_KEY).toBe('@aligent.cdk-utils.LambdaFunction');
        });

        it('should define context correctly', () => {
            // Arrange & Act
            const context = LambdaFunction.defineContext({
                memorySize: 256,
            });

            // Assert
            expect(context).toEqual({
                [LambdaFunction.CONTEXT_KEY]: {
                    memorySize: 256,
                },
            });
        });

        it('should have correct property injection ID', () => {
            // Arrange
            const mockLambdaFunction = Object.create(LambdaFunction.prototype);
            mockLambdaFunction.PROPERTY_INJECTION_ID = '@aligent.cdk-utils.LambdaFunction';

            // Assert
            expect(mockLambdaFunction.PROPERTY_INJECTION_ID).toBe(
                '@aligent.cdk-utils.LambdaFunction'
            );
        });
    });

    describe('integration tests', () => {
        let app: App;
        let stack: Stack;
        let template: Template;

        beforeEach(() => {
            app = new App();
            stack = new Stack(app, 'TestStack');
        });

        const synth = () => Template.fromStack(stack);

        test('creates a lambda function with default properties', () => {
            new LambdaFunction(stack, 'MyFunction', {
                entry: path.join(__dirname, '__data__', 'test-handler.ts'),
            });

            template = synth();
            const functions = template.findResources('AWS::Lambda::Function');
            expect(functions).toMatchSnapshot();
        });

        test('creates a lambda function with overridden properties via context', () => {
            const contextApp = new App({
                context: LambdaFunction.defineContext({ memorySize: 512 }),
            });
            const contextStack = new Stack(contextApp, 'TestStack');

            new LambdaFunction(contextStack, 'MyFunction', {
                entry: path.join(__dirname, '__data__', 'test-handler.ts'),
            });

            template = Template.fromStack(contextStack);
            const functions = template.findResources('AWS::Lambda::Function');
            expect(functions).toMatchSnapshot();
        });

        test('creates a lambda function with staging stage configuration', () => {
            const app = new App();
            const stagingStage = new StagingApplicationStage(app, {
                env: { account: '123456789012', region: 'us-east-1' },
            });
            const stagingStack = new Stack(stagingStage, 'TestStack');

            new LambdaFunction(stagingStack, 'MyFunction', {
                entry: path.join(__dirname, '__data__', 'test-handler.ts'),
            });

            template = Template.fromStack(stagingStack);

            // Check that staging configuration is applied
            template.hasResourceProperties('AWS::Logs::LogGroup', {
                RetentionInDays: 180,
            });

            template.hasResource('AWS::Logs::LogGroup', {
                DeletionPolicy: 'Delete',
            });

            template.hasResourceProperties('AWS::Lambda::Function', {
                Environment: {
                    Variables: {
                        NODE_OPTIONS: '--enable-source-maps',
                    },
                },
            });
        });

        test('creates a lambda function with development stage configuration', () => {
            const app = new App();
            const devStage = new DevelopmentApplicationStage(app, {
                env: { account: '123456789012', region: 'us-east-1' },
            });
            const devStack = new Stack(devStage, 'TestStack');

            new LambdaFunction(devStack, 'MyFunction', {
                entry: path.join(__dirname, '__data__', 'test-handler.ts'),
            });

            template = Template.fromStack(devStack);

            // Check that dev configuration is applied
            template.hasResourceProperties('AWS::Logs::LogGroup', {
                RetentionInDays: 7,
            });

            template.hasResource('AWS::Logs::LogGroup', {
                DeletionPolicy: 'Delete',
            });

            template.hasResourceProperties('AWS::Lambda::Function', {
                Environment: {
                    Variables: {
                        NODE_OPTIONS: Match.absent(),
                    },
                },
            });
        });

        test('creates a lambda function with alias', () => {
            new LambdaFunction(stack, 'MyFunction', {
                entry: path.join(__dirname, '__data__', 'test-handler.ts'),
                alias: 'LATEST',
            });

            template = synth();

            // Check that alias is created
            template.hasResourceProperties('AWS::Lambda::Alias', {
                Name: 'LATEST',
            });
        });

        test('throws error when not used within a CDK Stage (or Stack)', () => {
            const app = new App();

            expect(() => {
                new LambdaFunction(app, 'MyFunction', {
                    entry: path.join(__dirname, '__data__', 'test-handler.ts'),
                });
            }).toThrow('This construct must be used within a CDK Stage');
        });

        test('merges bundling properties correctly', () => {
            const app = new App({
                context: LambdaFunction.defineContext({
                    bundling: {
                        sourceMap: true,
                    },
                }),
            });
            const devStage = new DevelopmentApplicationStage(app, {
                env: { account: '123456789012', region: 'us-east-1' },
            });
            const devStack = new Stack(devStage, 'TestStack');

            // This test verifies that bundling properties are merged correctly
            // by checking that the function can be created without errors
            // and that source maps enable NODE_OPTIONS environment variable
            const lambdaFunction = new LambdaFunction(devStack, 'MyFunction', {
                entry: path.join(__dirname, '__data__', 'test-handler.ts'),
                bundling: {
                    minify: false,
                },
            });

            // Verify the function was created successfully
            expect(lambdaFunction).toBeDefined();

            template = Template.fromStack(devStack);

            // Check that NODE_OPTIONS is set due to sourceMap: true from context
            template.hasResourceProperties('AWS::Lambda::Function', {
                Environment: {
                    Variables: {
                        NODE_OPTIONS: '--enable-source-maps',
                    },
                },
            });
        });

        test('uses provided log group when specified', () => {
            const customLogGroup = new LogGroup(stack, 'CustomLogGroup');

            new LambdaFunction(stack, 'MyFunction', {
                entry: path.join(__dirname, '__data__', 'test-handler.ts'),
                logGroup: customLogGroup,
            });

            template = synth();

            // Should have the custom log group
            template.hasResourceProperties('AWS::Logs::LogGroup', {});

            // Should not create an additional log group with the default naming
            const logGroups = template.findResources('AWS::Logs::LogGroup');
            expect(Object.keys(logGroups).length).toBe(1);
        });
    });
});
