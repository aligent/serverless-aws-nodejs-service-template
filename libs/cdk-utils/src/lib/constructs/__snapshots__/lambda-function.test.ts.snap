// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LambdaFunction > integration tests > creates a lambda function with default properties 1`] = `
{
  "MyFunction3BAA72D1": {
    "DependsOn": [
      "MyFunctionServiceRole3C357FF2",
    ],
    "Properties": {
      "Code": {
        "S3Bucket": "[SNAPSHOT_PLACEHOLDER]",
        "S3Key": "[SNAPSHOT_PLACEHOLDER]",
      },
      "Environment": {
        "Variables": {
          "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          "NODE_OPTIONS": "--enable-source-maps",
        },
      },
      "Handler": "index.handler",
      "LoggingConfig": {
        "LogGroup": {
          "Ref": "awslambdaMyFunction9F5970AC",
        },
      },
      "Role": {
        "Fn::GetAtt": [
          "MyFunctionServiceRole3C357FF2",
          "Arn",
        ],
      },
      "Runtime": "nodejs16.x",
    },
    "Type": "AWS::Lambda::Function",
  },
}
`;

exports[`LambdaFunction > integration tests > creates a lambda function with overridden properties via context 1`] = `
{
  "MyFunction3BAA72D1": {
    "DependsOn": [
      "MyFunctionServiceRole3C357FF2",
    ],
    "Properties": {
      "Code": {
        "S3Bucket": "[SNAPSHOT_PLACEHOLDER]",
        "S3Key": "[SNAPSHOT_PLACEHOLDER]",
      },
      "Environment": {
        "Variables": {
          "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          "NODE_OPTIONS": "--enable-source-maps",
        },
      },
      "Handler": "index.handler",
      "LoggingConfig": {
        "LogGroup": {
          "Ref": "awslambdaMyFunction9F5970AC",
        },
      },
      "MemorySize": 512,
      "Role": {
        "Fn::GetAtt": [
          "MyFunctionServiceRole3C357FF2",
          "Arn",
        ],
      },
      "Runtime": "nodejs16.x",
    },
    "Type": "AWS::Lambda::Function",
  },
}
`;
